app/models/container_label_tag_mapping.rb

```diff
@@ -59,41 +66,33 @@ def self.tags_for_name_type_value(name, type, value)
       specific_value
     else
       any_value.map do |category_tag|
-        create_specific_value_mapping(name, type, value, category_tag).tag
+        specific_value_tag(name, value, category_tag)
       end
     end
   end
   private_class_method :tags_for_name_type_value
 
   # If this is an open ended any-value mapping, finds or creates a
   # specific-value mapping to a specific tag.
-  def self.create_specific_value_mapping(name, type, value, category_tag)
-    new_tag = create_tag(name, value, category_tag)
-    new_mapping = create!(:labeled_resource_type => type, :label_name => name, :label_value => value,
-                          :tag => new_tag)
-    load_mapping_into_hash(new_mapping)
-    new_mapping
-  end
-  private_class_method :create_specific_value_mapping
-
-  def self.create_tag(name, value, category_tag)
+  def self.specific_value_tag(name, value, category_tag)
     category = category_tag.classification
-    unless category
-      category = Classification.create_category!(:description => "Kubernetes label '#{name}'",
-                                                 :read_only   => true,
-                                                 :tag         => category_tag)
-    end
 
+    # Note: the names chosen here should remain stable,
+    # or we won't be able to find previously created tags.
     if value.empty?
       entry_name = ':empty:' # ':' character won't occur in kubernetes values.
       description = '<empty value>'
     else
       entry_name = Classification.sanitize_name(value)
       description = value
     end
-    entry = category.add_entry(:name => entry_name, :description => description)
-    entry.save!
+
+    entry = category.find_entry_by_name(entry_name)
+    unless entry
+      entry = category.add_entry(:name => entry_name, :description => description)
+      entry.save!  # TODO can this error?
```

Comment:
Hmm our whole codebase has just **one** lock usage (`MiqDatabase.with_lock`).
Plus I don't see how to use single-row locking for inserting a new row, and locking a whole table sound like definite overkill.

@kbrock @jrafanie Can you advise?  We want to upsert a Classification + Tag.  If 2 workers do it at once, they'll insert same data(*), there is no question who "wins", we just don't want duplicates.

(*) `.description` might have minor differences but we won't care.